obj: introduce customizable allocation header support

pbalcer opened the pull request:
This is the second and final patch related to: pmem/issues#347

This pull request also contains #1435 and #1366 because it depends
on those changes.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1533)
<!-- Reviewable:end -->

krzycz commented:
A few general comments:

1) "obj: _simplify_ memblock abstraction layer"
Really? Looks like we have different understanding on what "simplify" means. ;-)

2) The functions/variable names in some files are very long (i.e. 40 chars), which makes it hard to format the code nicely. Self-descriptive type/var names are welcome, but if the single assignment (very_long_name_var1 = very_long_name_var2) can't fit an 80-chars, there must be something wrong there...

Example: `container_seglists_insert_block()`, `container_seglists_get_rm_block_bestfit()`, `struct block_container_seglists`
Why not just `container_seglist_insert()` or `cntnr_sglist_insert()` or simply `seglist_insert()`?  And if it comes from the files name, then... rename the file to "seglist.c".
Also, could the structure name(s) be shortened somehow? I.e. blk_ctnr_sglist, bc_seglist, ...

---

Reviewed 120 of 126 files at r1, 6 of 6 files at r2, 8 of 8 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 26 unresolved discussions.

---

*[src/libpmemobj/alloc_class.h, line 65 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNPCb_PiGcs9uK8shp:-KaNPCb_PiGcs9uK8shq:b8box2e) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/alloc_class.h#L65)):*
> ```objc
> 
> 	union {
> 		/* struct { ... } huge; */
> ```

placeholder ? "XXX"?

---

*[src/libpmemobj/container_ctree.c, line 2 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNOoKDw2MlQORX-uDL:-KaNOoKDw2MlQORX-uDM:bb6ht2h) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/container_ctree.c#L2)):*
> ```C
> /*
>  * Copyright 2015-2016, Intel Corporation
> ```

2017

---

*[src/libpmemobj/container_ctree.c, line 76 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNOtGvcIeyNvk8t-nl:-KaNOtGvcIeyNvk8t-nm:bcy56bk) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/container_ctree.c#L76)):*
> ```C
> 	/*
> 	 * Even though the memory block representation of an object uses
> 	 * relatively large types in practise the entire memory block structure
> ```

practice

---

*[src/libpmemobj/container_seglists.c, line 34 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMoW4ypCdxOfze_8GE:-KaMoW4ypCdxOfze_8GF:b-g1g1ro) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/libpmemobj/container_seglists.c#L34)):*
> ```C
> 
> /*
>  * container_seglists.c -- implementation of segregated lists block container
> ```

Interesting. 3 years ago we developed almost exactly the same data structure (for the same purpose) for the PMEM allocator used in in the very first prototype of PMEM-aware Redis. :-)

---

*[src/libpmemobj/container_seglists.c, line 36 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMni4attuh1amplh3j:-KaMni4attuh1amplh3k:b-dvlses) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/libpmemobj/container_seglists.c#L36)):*
> ```C
>  * container_seglists.c -- implementation of segregated lists block container
>  *
>  * This container is constructed from N instrusive lists and a single 8 byte
> ```

N or 64 (or "up to 64")?
"instrusive"?

---

*[src/libpmemobj/container_seglists.c, line 212 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMntRMPc1ZpLHqXYK6:-KaMntRMPc1ZpLHqXYK7:b-9yyohp) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/libpmemobj/container_seglists.c#L212)):*
> ```C
> 	bc->super.c_ops = &container_seglists_ops;
> 
> 	for (uint32_t i = 0; i < SEGLIST_BLOCK_LISTS; ++i)
> ```

int or unsigned?

---

*[src/libpmemobj/ctree.c, line 395 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNOhj6rM0cx83-FRJC:-KaNOhj6rM0cx83-FRJD:bpcgj8k) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/ctree.c#L395)):*
> ```C
> 	}
> 
> 	struct node_leaf *l = *dst;
> ```

l => leaf

---

*[src/libpmemobj/memblock.c, line 162 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaarA5flQbWcCdY6QCe:-KaarA5flQbWcCdY6QCf:b-i961cv) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/memblock.c#L162)):*
> ```C
> 	struct allocation_header_legacy *hdr = m->m_ops->get_real_data(m);
> 
> 	return (uint16_t)(hdr->root_size >> 48ULL);
> ```

magic number

---

*[src/libpmemobj/obj.c, line 684 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNRmQ8qAfVWUaVJsOw:-KaNRmQ8qAfVWUaVJsOx:bwxwwhg) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/obj.c#L684)):*
> ```C
> 	pmemops_persist(p_ops, &pop->root_offset, sizeof(pop->root_offset));
> 	pop->root_size = 0;
> 	pmemops_persist(p_ops, &pop->root_size, sizeof(pop->root_size));
> ```

1) Good opportunity to move initialization of those fields (run_id, root_offset, root_size) to the end. This would make it clear they are not in the persistent (immutable) part of the descriptor.
2) Seems like it might be a good idea to zero out (and flush) the entire descriptor (4K) first, not only 2K (OBJ_DSC_P_SIZE) -
pmemops_memset_persist(). Then, we don't need to initialize mentioned fields at all.

---

*[src/libpmemobj/palloc.c, line 165 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNVkBXZ9VNEIi7oHOx:-KaNVkBXZ9VNEIi7oHOy:b2nhwoh) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/palloc.c#L165)):*
> ```C
> 	 * To provide optimal scaling for multi-threaded applications and reduce
> 	 * fragmentation the appropriate bucket is chosen depending on the
> 	 * current thread context and to which allocation class the requested*
> ```

`*`

---

*[src/libpmemobj/palloc.h, line 51 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaaryVV0tFLTv24GMlx:-KaaryVWutApBASEvOAb:bgo5tou) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/palloc.h#L51)):*
> ```objc
> 	struct heap_rt *rt;
> 	uint64_t size;
> 	uint64_t run_id;
> ```

What is the purpose of this field?

---

*[src/libpmemobj/recycler.c, line 2 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-Kaas3Ih8o1SwpVeIfEZ:-Kaas3Ih8o1SwpVeIfE_:bb6ht2h) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/recycler.c#L2)):*
> ```C
> /*
>  * Copyright 2016, Intel Corporation
> ```

2017

---

*[src/libpmemobj/recycler.c, line 47 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaasAwqXxjMfmqXbZW1:-KaasAwqXxjMfmqXbZW2:b-xu7pyd) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/recycler.c#L47)):*
> ```C
> 
> #define RUN_KEY_GET_ZONE_ID(k)\
> ((uint16_t)(k))
> ```

Even though it's correct, it looks strange w/o "& 0xFFFF".

---

*[src/libpmemobj/recycler.c, line 108 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaasvA-vh8QDXLmCDR6:-KaasvA-vh8QDXLmCDR7:bbpeit9) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/recycler.c#L108)):*
> ```C
> 	for (int i = 0; i < MAX_BITMAP_VALUES; ++i)
> 		if (run->bitmap[i] != UINT64_MAX)
> 			score += !__builtin_popcountll(run->bitmap[i]);
> ```

Does it compile on Windows?

---

*[src/libpmemobj/recycler.h, line 36 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-Kaat7rE1YpyOz50k9Y_:-Kaat7rE1YpyOz50k9Ya:b8korqb) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/recycler.h#L36)):*
> ```objc
>  * recycler.h -- internal definitions of run recycler
>  *
>  * This is a container that stores runs that are currently not used by any of
> ```

Does it means it is some sort of a "shared" bucket (with respect to threads)?

---

*[src/test/obj_constructor/TEST2, line 3 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaN81hVsP2b4WO8fh1X:-KaN81hVsP2b4WO8fh1Y:bb6ht2h) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_constructor/TEST2#L3)):*
> ```
> #!/bin/bash -e
> #
> # Copyright 2015-2016, Intel Corporation
> ```

2017

---

*[src/test/obj_first_next/out0.log.match, line 33 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNCEteYzvwP31h_fGA:-KaNCEteYzvwP31h_fGB:b-qhk1i1) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_first_next/out0.log.match#L33)):*
> ```
> constructor(id = 4)
> type:
> id = $(nW)
> ```

There is no point in printing any *id's* if we ignore the values.

---

*[src/test/obj_heap/obj_heap.c, line 74 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNMGSH7YVybjvwsBRb:-KaNMGSH7YVybjvwsBRc:bdeqpal) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_heap/obj_heap.c#L74)):*
> ```C
> 
> static void
> init_run_with_score(struct heap_layout *l, uint32_t chunk_id, int score)
> ```

"l" => "hl" ?

---

*[src/test/obj_heap/obj_heap.c, line 84 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNM3jPCUU42KZK7dkk:-KaNM3jPCUU42KZK7dkl:b-nivchc) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_heap/obj_heap.c#L84)):*
> ```C
> 	run->block_size = 1024;
> 	memset(run->bitmap, 0xFF, sizeof(run->bitmap));
> 	UT_ASSERT(score % 64 == 0);
> ```

UT_ASSERTeq / UT_ASSERTne (here and below)

---

*[src/test/obj_list_valgrind/TEST4, line 47 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMpOveEcMnMiyTh_pr:-KaMpOveEcMnMiyTh_ps:b3kre) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/obj_list_valgrind/TEST4#L47)):*
> ```
> 
> require_fs_type pmem non-pmem
> #configure_valgrind pmemcheck force-enable ../obj_list/obj_list
> ```

?

---

*[src/test/obj_pmalloc_basic/Makefile, line 40 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNN_WiM2RDrNs9aOS6:-KaNN_WiM2RDrNs9aOS7:b3kre) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_pmalloc_basic/Makefile#L40)):*
> ```Makefile
> 
> LIBPMEM=y
> LIBPMEMOBJ=internal-debug
> ```

?

---

*[src/test/obj_pmalloc_basic/obj_pmalloc_basic.c, line 192 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNNGSRR5-7J90-Jx3I:-KaNNGSRR5-7J90-Jx3J:b-i961cv) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_pmalloc_basic/obj_pmalloc_basic.c#L192)):*
> ```C
> 	UT_ASSERTeq(palloc_extra(&pop->heap, val), PMALLOC_EXTRA);
> 	UT_ASSERT((palloc_flags(&pop->heap, val) & PALLOC_FLAG) == PALLOC_FLAG);
> 	UT_ASSERT(palloc_usable_size(&pop->heap, val) == 112);
> ```

magic number

---

*[src/test/obj_realloc/Makefile, line 41 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNNX0KAfyNdN7bd-KB:-KaNNX0KAfyNdN7bd-KC:b3kre) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_realloc/Makefile#L41)):*
> ```Makefile
> LIBPMEMCOMMON=y
> LIBPMEM=y
> LIBPMEMOBJ=internal-debug
> ```

?

---

*[src/test/obj_rpmem_basic_integration/node_0_out5.log.match, line 8 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMXONVqLtjvb9MU2Ow:-KaMXONWyjdGpRUWWxtV:b-7ztj2c) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/obj_rpmem_basic_integration/node_0_out5.log.match#L8)):*
> ```
> free
> realloc: 0 => 777, size: $(N)
> realloc: 777 => 1, size: $(N)
> ```

Well, now this test does not check if realloc works as expected. I.e. you don't know if in case of reallocation from size 777  to 1, the object is really downsized. (there was a time it didn't work - that's why this case was added to the test).
Perhaps, instead of using match file, we should move this logic to the test and do something like this:
ASSERTeq(actual_size, ALIGN(requested_size, size_to_alloc_class(requested_size)));

---

*[src/test/tools/pmemalloc/Makefile, line 43 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaN6pBOgyoXffXNdsnV:-KaN6pBOgyoXffXNdsnW:b3kre) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/tools/pmemalloc/Makefile#L43)):*
> ```Makefile
> LIBPMEMOBJ=y
> 
> CFLAGS += -I$(TOP)/src/libpmemobj -g
> ```

?

---

*[src/test/tools/pmemobjcli/Makefile, line 46 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaN6leXLPccbukvjvey:-KaN6leXLPccbukvjvez:b3kre) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/tools/pmemobjcli/Makefile#L46)):*
> ```Makefile
> TOOLS_COMMON=y
> LIBPMEMBLK_PRIV=btt_info_convert2h
> CFLAGS+=-g -O0
> ```

?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1533#-:-KaauGZvJqL2dTZbtHVv:brlmedx)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
1. That specific patch removed the MEMBLOCK_OPS set of macros and traded off a little bit of transient storage for smaller CPU footprint and easier to use API.
In the context of the entire patch, the memblock abstraction layer got less leaky and the heap.c shared code is now much less intertwined with everything else.
2. I agree, I'll think about changing some of the names.

---

Review status: all files reviewed at latest revision, 26 unresolved discussions.

---

*[src/libpmemobj/alloc_class.h, line 65 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNPCb_PiGcs9uK8shp:-Kam5Rk-Fe56RbXBZsHs:b-z7to17) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/alloc_class.h#L65)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

placeholder ? "XXX"?
</blockquote></details>

this is a union with only a single structure. I've initially had a second, empty structure, but vs complained. This is constructed this way to indicate the way this structure should evolve.

---

*[src/libpmemobj/container_ctree.c, line 2 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNOoKDw2MlQORX-uDL:-Kam5pygyDxxhZ-xg-4s:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/container_ctree.c#L2)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

2017
</blockquote></details>

Done.

---

*[src/libpmemobj/container_ctree.c, line 76 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNOtGvcIeyNvk8t-nl:-Kam5qPOaMPmbA_h59sF:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/container_ctree.c#L76)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

practice
</blockquote></details>

Done.

---

*[src/libpmemobj/container_seglists.c, line 36 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMni4attuh1amplh3j:-Kam5thVRqeVGcEcPCzW:bw29pyv) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/libpmemobj/container_seglists.c#L36)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

N or 64 (or "up to 64")?
"instrusive"?
</blockquote></details>

N as in "you can choose whichever number of lists you want". I've added up to 64.

An intrusive data structure is one that requires its elements to have knowledge about the container.

---

*[src/libpmemobj/container_seglists.c, line 212 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMntRMPc1ZpLHqXYK6:-Kam6a4mz0K1HWB9PVz6:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/libpmemobj/container_seglists.c#L212)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

int or unsigned?
</blockquote></details>

Done.

---

*[src/libpmemobj/ctree.c, line 395 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNOhj6rM0cx83-FRJC:-Kam6fJfTaLe--2LUf9P:b1ya000) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/ctree.c#L395)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

l => leaf
</blockquote></details>

I agree with making the variables a bit more verbose, but I don't want to change a single variable. I'll refactor the entire ctree code soonish to make it a bit prettier.

---

*[src/libpmemobj/memblock.c, line 162 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaarA5flQbWcCdY6QCe:-Kam6tB6_tdoTDpcNsXg:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/memblock.c#L162)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

magic number
</blockquote></details>

Done.

---

*[src/libpmemobj/palloc.c, line 165 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNVkBXZ9VNEIi7oHOx:-Kam6z9b350PezTfPs3k:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/palloc.c#L165)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

`*`
</blockquote></details>

Done.

---

*[src/libpmemobj/palloc.h, line 51 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaaryVV0tFLTv24GMlx:-Kam6ztUwJHl6KKMRyof:b-8wg3z7) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/palloc.h#L51)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

What is the purpose of this field?
</blockquote></details>

I needed a way to determine whether a chunk is 'owned' by any of the threads in the current incarnation of the heap. Each time a thread takes ownership of a chunk, it stores a run_id metadata and similarly when a thread no longer needs a chunk, it zeroes the relevant field.

---

*[src/libpmemobj/recycler.c, line 2 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-Kaas3Ih8o1SwpVeIfEZ:-Kam7TrJjOV1-jn9f0Dm:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/recycler.c#L2)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

2017
</blockquote></details>

Done.

---

*[src/libpmemobj/recycler.c, line 108 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaasvA-vh8QDXLmCDR6:-Kam7dEQ1-N2l0zyoJwN:b-tboj2w) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/recycler.c#L108)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Does it compile on Windows?
</blockquote></details>

Yes, this entire patch compiles on windows :)

---

*[src/libpmemobj/recycler.h, line 36 at r4](https://reviewable.io:443/reviews/pmem/nvml/1533#-Kaat7rE1YpyOz50k9Y_:-Kam7g1BpLpmh5AvG82_:bc8r87) ([raw file](https://github.com/pmem/nvml/blob/afd7d7f947ba30ad5cb875b2f50a657b5b04d01d/src/libpmemobj/recycler.h#L36)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Does it means it is some sort of a "shared" bucket (with respect to threads)?
</blockquote></details>

Yes, in fact initially I wanted to simply create a 'struct bucket recycler;' and use that, but it became obvious that I would have no way of accurately 'scoring' the runs without some intermediary layer. This is cleaner.

---

*[src/test/obj_first_next/out0.log.match, line 33 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNCEteYzvwP31h_fGA:-Kam83hDdupgviPBVXNF:bcidrn9) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_first_next/out0.log.match#L33)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

There is no point in printing any *id's* if we ignore the values.
</blockquote></details>

We already have different tests that check correctness of first/next functions in a programmatic manner. The output of this test is dependent on the internal heap layout, and is annoying to deal with. For example, the id's can be in a different order if you change the object sizes...

---

*[src/test/obj_heap/obj_heap.c, line 74 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNMGSH7YVybjvwsBRb:-Kam8fR3_JOa13TuR06I:bfad7ct) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_heap/obj_heap.c#L74)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

"l" => "hl" ?
</blockquote></details>

"l" is consistent with other usages of this structure.

---

*[src/test/obj_heap/obj_heap.c, line 84 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNM3jPCUU42KZK7dkk:-Kam8r3tzrgsHR07Bc4O:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_heap/obj_heap.c#L84)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

UT_ASSERTeq / UT_ASSERTne (here and below)
</blockquote></details>

Done.

---

*[src/test/obj_list_valgrind/TEST4, line 47 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMpOveEcMnMiyTh_pr:-Kam8rilo1HoXbRRjIss:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/obj_list_valgrind/TEST4#L47)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

?
</blockquote></details>

Done.

---

*[src/test/obj_pmalloc_basic/Makefile, line 40 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNN_WiM2RDrNs9aOS6:-Kam8seY93RaEgNswe3z:blbs9w6) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_pmalloc_basic/Makefile#L40)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

?
</blockquote></details>

This should be linked with debug version of the library.

---

*[src/test/obj_pmalloc_basic/obj_pmalloc_basic.c, line 192 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNNGSRR5-7J90-Jx3I:-Kam8x1O_C9i2Uw5mIAS:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_pmalloc_basic/obj_pmalloc_basic.c#L192)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

magic number
</blockquote></details>

Done.

---

*[src/test/obj_realloc/Makefile, line 41 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNNX0KAfyNdN7bd-KB:-Kam8xiv7HfbwdyF9xnn:b-9y5vgb) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_realloc/Makefile#L41)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

?
</blockquote></details>

I've changed this test to use real allocation classes using code from the library, hence requirement to link with it.

---

*[src/test/obj_rpmem_basic_integration/node_0_out5.log.match, line 8 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMXONVqLtjvb9MU2Ow:-Kam960E7j-oF9RVIxGq:b-luuuwf) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/obj_rpmem_basic_integration/node_0_out5.log.match#L8)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Well, now this test does not check if realloc works as expected. I.e. you don't know if in case of reallocation from size 777  to 1, the object is really downsized. (there was a time it didn't work - that's why this case was added to the test).
Perhaps, instead of using match file, we should move this logic to the test and do something like this:
ASSERTeq(actual_size, ALIGN(requested_size, size_to_alloc_class(requested_size)));

</blockquote></details>

obj_realloc tests whether or not realloc actually works with comprehensive programmatic checks.
Hard-codding numbers into match files makes it more difficult to change anything.

---

*[src/test/tools/pmemalloc/Makefile, line 43 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaN6pBOgyoXffXNdsnV:-Kam9RDdE0BYhTkQjYd9:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/tools/pmemalloc/Makefile#L43)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

?
</blockquote></details>

Done.

---

*[src/test/tools/pmemobjcli/Makefile, line 46 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaN6leXLPccbukvjvey:-Kam9Rba4XayZ4JHNT0r:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/tools/pmemobjcli/Makefile#L46)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1533#-:-Kam9SjkB4R7SpEndmyn:bygnkr9)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 10 of 10 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks broke.

---

*[src/libpmemobj/container_seglists.c, line 36 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMni4attuh1amplh3j:-KaqoKRZD3TOiGNhF0k3:b-xzgvwb) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/libpmemobj/container_seglists.c#L36)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

N as in "you can choose whichever number of lists you want". I've added up to 64.

An intrusive data structure is one that requires its elements to have knowledge about the container.
</blockquote></details>

typo: in_s_trusive

---

*[src/libpmemobj/ctree.c, line 395 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNOhj6rM0cx83-FRJC:-KaqnytUO6krpz01itTe:bie2x42) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/ctree.c#L395)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

I agree with making the variables a bit more verbose, but I don't want to change a single variable. I'll refactor the entire ctree code soonish to make it a bit prettier.
</blockquote></details>

It's not about making it verbose - it's about not using "l" as var name. It could be "nl" (node leaf)...

---

*[src/test/obj_first_next/out0.log.match, line 33 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNCEteYzvwP31h_fGA:-KbAOaU9eRtz9tu5FUBl:b-cpw5sk) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_first_next/out0.log.match#L33)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

We already have different tests that check correctness of first/next functions in a programmatic manner. The output of this test is dependent on the internal heap layout, and is annoying to deal with. For example, the id's can be in a different order if you change the object sizes...
</blockquote></details>

So, again - it looks like there is no point in printing ID's anymore if the order can change.
Maybe we could just check the number of objects on given list and/or check if all the expected ID's are still there (i.e. sum of ID's should be still the same, even if the order changes, so this could be printed instead).

---

*[src/test/obj_pmalloc_basic/Makefile, line 40 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNN_WiM2RDrNs9aOS6:-KbAQHa-8BayTq9yILeF:b4rsm8o) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_pmalloc_basic/Makefile#L40)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

This should be linked with debug version of the library.
</blockquote></details>

Why? I don't see anything in this test that would require debug version of the library.

---

*[src/test/obj_pmalloc_basic/obj_pmalloc_basic.c, line 192 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNNGSRR5-7J90-Jx3I:-KaqIiNKOkZtMNnKzFsm:bvkvbq5) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_pmalloc_basic/obj_pmalloc_basic.c#L192)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

It still doesn't explain why it's 112, not 124 or 98. I thought it came from some calculation / metadata size.

---

*[src/test/obj_rpmem_basic_integration/node_0_out5.log.match, line 8 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMXONVqLtjvb9MU2Ow:-KbAHK6tnuPVY7GoN8SJ:bwwjqqw) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/obj_rpmem_basic_integration/node_0_out5.log.match#L8)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

obj_realloc tests whether or not realloc actually works with comprehensive programmatic checks.
Hard-codding numbers into match files makes it more difficult to change anything.
</blockquote></details>

1) Maybe I'm wrong, but when I look at obj_realloc test, I can't see the code that checks whether realloc to smaller size really shrinks the object. It only checks if the usable sizes are not less than requested sizes (before and after realloc).

2) Assuming obj_realloc test covers that - if the allocation sizes are to be ignored in the match file, there is no need to print them at all in *_basic_integration tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1533)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: 83 of 132 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/libpmemobj/container_seglists.c, line 36 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMni4attuh1amplh3j:-KbKzJ3UK1ilT3e2Mk2q:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/libpmemobj/container_seglists.c#L36)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

typo: in_s_trusive
</blockquote></details>

Done.

---

*[src/libpmemobj/ctree.c, line 395 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNOhj6rM0cx83-FRJC:-KbKzKWDAAK1fuRwJbWF:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/ctree.c#L395)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

It's not about making it verbose - it's about not using "l" as var name. It could be "nl" (node leaf)...
</blockquote></details>

Done.

---

*[src/libpmemobj/obj.c, line 684 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNRmQ8qAfVWUaVJsOw:-KbKzLgh-_Ql-eLjsFD3:btmgnnx) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/libpmemobj/obj.c#L684)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

1) Good opportunity to move initialization of those fields (run_id, root_offset, root_size) to the end. This would make it clear they are not in the persistent (immutable) part of the descriptor.
2) Seems like it might be a good idea to zero out (and flush) the entire descriptor (4K) first, not only 2K (OBJ_DSC_P_SIZE) -
pmemops_memset_persist(). Then, we don't need to initialize mentioned fields at all.
</blockquote></details>

I've moved initialization of those fields to the bottom of the function. The second part can be done as a separate patch (I don't want to make this one bigger).

---

*[src/test/obj_first_next/out0.log.match, line 33 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNCEteYzvwP31h_fGA:-KbKzfED2QHPeT8FJnjs:byxb4mi) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_first_next/out0.log.match#L33)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

So, again - it looks like there is no point in printing ID's anymore if the order can change.
Maybe we could just check the number of objects on given list and/or check if all the expected ID's are still there (i.e. sum of ID's should be still the same, even if the order changes, so this could be printed instead).
</blockquote></details>

The code for this test is structured in a way that makes your suggestion quite difficult to actually implement.

---

*[src/test/obj_pmalloc_basic/Makefile, line 40 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNN_WiM2RDrNs9aOS6:-KbL-ZRm-eZi5YkQolQd:bgabdmm) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_pmalloc_basic/Makefile#L40)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Why? I don't see anything in this test that would require debug version of the library.
</blockquote></details>

By this logic you could say that there is no point in compiling any tests with the debug version of the library because it is not technically speaking required...

This is a test that I heavily rely on during changes of the allocator and I need it to be compiled with debug version for debugging.

---

*[src/test/obj_pmalloc_basic/obj_pmalloc_basic.c, line 192 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNNGSRR5-7J90-Jx3I:-KbL-xxIZ8Zt79uggs-0:b-ou54he) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_pmalloc_basic/obj_pmalloc_basic.c#L192)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

It still doesn't explain why it's 112, not 124 or 98. I thought it came from some calculation / metadata size.
</blockquote></details>

"FIRST_SIZE" as in the first class size.

---

*[src/test/obj_rpmem_basic_integration/node_0_out5.log.match, line 8 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMXONVqLtjvb9MU2Ow:-KbL08A0umkIlX8FzN2h:b-zd72q7) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/obj_rpmem_basic_integration/node_0_out5.log.match#L8)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

1) Maybe I'm wrong, but when I look at obj_realloc test, I can't see the code that checks whether realloc to smaller size really shrinks the object. It only checks if the usable sizes are not less than requested sizes (before and after realloc).

2) Assuming obj_realloc test covers that - if the allocation sizes are to be ignored in the match file, there is no need to print them at all in *_basic_integration tests.
</blockquote></details>

Because technically speaking reallocation to a smaller size does not guarantee shrinking of the underlying buffer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1533)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 49 of 49 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

*[src/test/obj_pmalloc_basic/Makefile, line 40 at r3](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaNN_WiM2RDrNs9aOS6:-KbLSJBBQkeeQq_LTVoc:bwfcw8g) ([raw file](https://github.com/pmem/nvml/blob/a198af8db641cb116748b2d62d38ee9db8143e36/src/test/obj_pmalloc_basic/Makefile#L40)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

By this logic you could say that there is no point in compiling any tests with the debug version of the library because it is not technically speaking required...

This is a test that I heavily rely on during changes of the allocator and I need it to be compiled with debug version for debugging.
</blockquote></details>

OK, now I see what is the problem.
Perhaps we should use internal-debug/internal-nondebug depending on whether this is a static-debug or static-nondebug version of the test binary. I.e. use `LIBPMEMOBJ=internal` and let make (Makefile.inc) to select the proper variant.
Unfortunately, we only have one "dynamic" version of the test binary, so it must be either debug or nondebug (with respect to statically linked libs) unless we improve our makefiles.

---

*[src/test/obj_rpmem_basic_integration/node_0_out5.log.match, line 8 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMXONVqLtjvb9MU2Ow:-KbLVFNsHe0GRoMr9gnJ:bb7ne25) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/obj_rpmem_basic_integration/node_0_out5.log.match#L8)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Because technically speaking reallocation to a smaller size does not guarantee shrinking of the underlying buffer.
</blockquote></details>

True, but as a user, I could expect that doing reallocation i.e. from 4M to 4B, the object is actually shrinked - not exactly to 4B, of course, but something much less than 4M. IOW, to the smallest possible size class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1533)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 14 of 126 files at r1, 2 of 10 files at r5, 23 of 49 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[src/libpmemobj/alloc_class.c, line 135 at r6](https://reviewable.io:443/reviews/pmem/nvml/1533#-KbUsRGQmVETkZSt6RDs:-KbUsRGQmVETkZSt6RDt:boyvvlv) ([raw file](https://github.com/pmem/nvml/blob/a03af547e7059e58b3554dce2723fffcdf5178ca/src/libpmemobj/alloc_class.c#L135)):*
> ```C
> {
> 	int n;
> 	for (n = 0; n < MAX_ALLOCATION_CLASSES; ++n) {
> ```

int n = 0; ... ?

---

*[src/libpmemobj/alloc_class.c, line 341 at r6](https://reviewable.io:443/reviews/pmem/nvml/1533#-KbUt3gs4wwT4RN1Gxwu:-KbUt3gtECrHjWt4dSoq:b-w7l33v) ([raw file](https://github.com/pmem/nvml/blob/a03af547e7059e58b3554dce2723fffcdf5178ca/src/libpmemobj/alloc_class.c#L341)):*
> ```C
> 			RUN_UNIT_MAX, RUN_UNIT_MAX_ALLOC);
> 	if (predefined_class == NULL)
> 		goto error_alloc_class_create;
> ```

blank line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1533)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[src/libpmemobj/alloc_class.c, line 135 at r6](https://reviewable.io:443/reviews/pmem/nvml/1533#-KbUsRGQmVETkZSt6RDs:-KbjQXYiAsJ3x9w_jeLE:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a03af547e7059e58b3554dce2723fffcdf5178ca/src/libpmemobj/alloc_class.c#L135)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

int n = 0; ... ?
</blockquote></details>

Done.

---

*[src/libpmemobj/alloc_class.c, line 341 at r6](https://reviewable.io:443/reviews/pmem/nvml/1533#-KbUt3gs4wwT4RN1Gxwu:-KbjQY-YW9Yee6UtdVTA:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a03af547e7059e58b3554dce2723fffcdf5178ca/src/libpmemobj/alloc_class.c#L341)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

blank line
</blockquote></details>

Done.

---

*[src/test/obj_rpmem_basic_integration/node_0_out5.log.match, line 8 at r2](https://reviewable.io:443/reviews/pmem/nvml/1533#-KaMXONVqLtjvb9MU2Ow:-KbjQbPaJklY-J-pXht2:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/75619df763070437eae197ff5e5b0c6e23cb1d15/src/test/obj_rpmem_basic_integration/node_0_out5.log.match#L8)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

True, but as a user, I could expect that doing reallocation i.e. from 4M to 4B, the object is actually shrinked - not exactly to 4B, of course, but something much less than 4M. IOW, to the smallest possible size class.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1533)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 12 of 12 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1533)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1533#-:-KbkcXHb7GHu-YVoMSTO:bnfp4nl)*
<!-- Sent from Reviewable.io -->
